lms/services/canvas_api/README.md

```diff
@@ -0,0 +1,49 @@
+# Canvas API Client
+
+## The class layout
+
+There are a number of different API clients which deal with the API at varying
+different levels. They all sit inside the `public_client.CanvasAPIClient` in
+a nested fashion:
+
+ * [`public.CanvasAPIClient`](public.py) - High level API access
+ * [`authenticated.AuthenticatedClient`](authenticated.py) - OAuth2 token handling
+ * [`basic.BasicClient`](basic.py) - HTTP marshalling, requests and applying schema
+
+To see how these are all bolted together see: [`__init__.py`](__init__.py)
```

Comment:
I think this is a good layout and both works with how we've described this stuff before, and also goes to shore up a few fuzzy ideas I was having.

The first is that in one sense `canvas_api_client_factory()` is the interface to this code, except... not at all. When this is working correctly, nobody at all interacts with that function. Everybody who is talking to this code is interacting with `CanvasAPIClient`.

Another mistake is "technically" the interface is defined in `__init__.py`. So it's true to say that we could not have underscores and you "are informed" of the public and private interface. But we don't have to make it hard. What's exptected to be exported in the `__init__.py` should probably reflect what's not underscored in the directoy.

By having the factory in a separate method, we also get around having to use `__all__` to make the distinction between things we imported into `__init__.py` because they are needed locally, and things which are intended for export. If it's in `__init__.py` it's because it's part of the public interface (import loops permitting).

This also allows us to work out what we should do in a few other cases. Lets say we had something like:

```python
__init__.py
   from factory import factory_method()
   from client import Client Object
factory.py
    factory_method()
client.py
    ClientObject()
    ClientObjectInFileCollaborator()
_client_sub_object.py
    ClientSubObject()
    _ClientSubObjectInFileCollaborator()
_client_return_model.py
   ReturnModel()
```

There a number of mistakes here:

 * `ClientObjectInFileCollaborator` should be private, as it's not part of the interface, and it is in a file marked as part of the interface
 * `_ClientSubObjectInFileCollaborator` should not be private, as it's already marked private by being in a private marked file. 
* Assuming `ReturnModel` is something sent back when interacting with `ClientObject` then it forms part of the surface area you interact with when using the code. This makes it part of the public interface too, even if it's at the bottom of the call stack. So `_client_return_model.py` should be public too. 
 * This also means `__init__.py` should import `ReturnModel` to further advertise it's public-ness and cut of outsiders from the exact location of things within the context (allowing refactoring)